common/include/pcl/common/impl/eigen.hpp

```diff
@@ -248,6 +249,33 @@ pcl::computeCorrespondingEigenVector (const Matrix& mat, const typename Matrix::
     eigenvector = vec3 / std::sqrt (len3);
 }
 
+namespace pcl {
+namespace detail{
+template <typename Matrix, typename Scalar>
+struct EigenVector {
+  Matrix vector;
+  Scalar length;
+};
+
+template <typename Matrix> static EigenVector<Matrix, typename Matrix::Scalar>
+get_largest_3x3_eigenvector (const Matrix scaledMatrix)
+{
+  std::array<Vector, 3> vector {scaledMatrix.row (0).cross (scaledMatrix.row (1)),
+                                scaledMatrix.row (0).cross (scaledMatrix.row (2)),
+                                scaledMatrix.row (1).cross (scaledMatrix.row (2))};
+  std::array<Scalar, 3> len;
+  std::transform (vector.cbegin(), vector.cend (), len.begin (),
+      [](const auto& vec) { return vec.squaredNorm (); });
+
+  const auto index = std::distance (len.cbegin (),
+      std::max_element (len.cbegin (), len.cend ()));
+
+  const auto eigenvector = vector[index] / std::sqrt(len[index]);
+  return EigenVector{eigenvector, len[index]};
```

Comment:
I currently don't have time to implement what I'm gonna describe but I have a strong opinion this whole function should be totally implemented with eigen types. 

Construct your 3x3 matrix ([comma initializer](https://eigen.tuxfamily.org/dox/group__TutorialAdvancedInitialization.html)) from the cross product results. Perform a col-wise norm using the [`colwise()`](https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html#TutorialReductionsVisitorsBroadcastingPartialReductions) reduction and get the length/eigenvalue and vector with `maxCoeff()` extract both the value and coefficient.

Reply:
Something like this?